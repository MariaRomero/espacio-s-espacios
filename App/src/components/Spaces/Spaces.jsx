import React, { useEffect, useState } from "react";
import firebase from "../../firebase";
import Form from "../Form/Form";
import Space from "../Space/Space";
import styles from "./styles.module.css";
import Popup from "reactjs-popup";

function Spaces() {
  const [spaces, setSpaces] = useState();
  const [updatingSpace, setUpdatingSpace] = useState(null);

  useEffect(() => {
    const spacesRef = firebase.database().ref("spaces");
    spacesRef.on("value", snapshot => {
      let spacesObj = snapshot.val();
      let newSpaces = [];

      for (let space in spacesObj) {
        newSpaces.push({
          id: space,
          address: spacesObj[space].address,
          postcode: spacesObj[space].postcode,
          noBeds: spacesObj[space].noBeds,
          description: spacesObj[space].description,
          price: spacesObj[space].price,
          images: spacesObj[space].images,
          expires: spacesObj[space].expires
        });
      }

      setSpaces(newSpaces);
    });
  }, []);

  const removeSpace = spaceId => {
    const spaceRef = firebase.database().ref(`/spaces/${spaceId}`);
    spaceRef
      .remove()
      .then(() => {
        alert("Space successfully deleted!");
      })
      .catch(error => {
        console.error("Error removing space: ", error);
      });
  };

  function updateForm(event, space) {
    setUpdatingSpace(space);
  }

  return (
    <section className={styles.container}>
      <div className="wrapper">
        <ul className={styles.list}>
          {typeof spaces === "undefined"
            ? "Loading..."
            : spaces.map(space => {
                return (
                  <Space
                    key={space.id}
                    space={space}
                    removeSpace={removeSpace}
                    updateForm={updateForm}
                  />
                );
              })}
        </ul>
        {updatingSpace && (
          <Popup
            onClose={() => setUpdatingSpace(false)}
            open={updatingSpace}
            position="right center"
          >
            <div className={styles.popupContent}>
              <button
                className={styles.cancelEditingAction}
                onClick={() => setUpdatingSpace(false)}
              >
                Cancel
              </button>
              <Form
                spaceSelected={updatingSpace}
                setUpdatingSpace={setUpdatingSpace}
              />
            </div>
          </Popup>
        )}
      </div>
    </section>
  );
}

export default Spaces;
