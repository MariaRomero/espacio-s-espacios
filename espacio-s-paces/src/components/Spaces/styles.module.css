.container {
  padding: 8px;
  flex: initial;
  display: block;
}

@media (min-width: 680px) {
  .container {
    background-color: #fff;
    padding: 8px 8px 8px 32px;
    flex: 1 1 auto;
    margin: 0 auto;
    font-weight: 300;
  }
}

.list {
  display: flex;
  flex-wrap: wrap;
  margin: 0;
  padding: 0;
}

.cancelEditingAction {
  align-self: flex-end;
  color: transparent;
  cursor: pointer;
  border: 0;
  display: inline-block;
  padding: 10px 6px;
  width: 20px;
  height: 20px;
  background: linear-gradient(
      45deg,
      rgba(7, 53, 122, 0) 0%,
      rgba(7, 53, 122, 0) 43%,
      #43357a 45%,
      #43357a 55%,
      rgba(7, 53, 122, 0) 57%,
      rgba(67, 53, 122, 0) 100%
    ),
    linear-gradient(
      135deg,
      transparent 0%,
      transparent 43%,
      #43357a 45%,
      #43357a 55%,
      transparent 57%,
      transparent 100%
    );
}

.popupContent {
  display: flex;
  flex-direction: column;
}
