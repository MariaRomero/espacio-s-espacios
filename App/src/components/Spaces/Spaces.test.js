import "@testing-library/jest-dom/extend-expect";

import React from "react";
import { render } from "@testing-library/react";
import Spaces from "./Spaces";
import renderer from "react-test-renderer";

it("renders Spaces created correctly", () => {
  const component = renderer.create(<Spaces />).toJSON();
  expect(component).toMatchSnapshot();
});

test("renders loading when spaces is undefined", () => {
  const testData = "Loading...";
  const { getByText } = render(<Spaces />);

  expect(getByText(testData)).toBeInTheDocument();
});

test("renders components and displays text", () => {
  const spaceObj = [
    {
      id: "1234",
      address: "abc road",
      postcode: "bn12nb",
      noBeds: 2,
      description: "lovely house with garden",
      price: 230,
      images: "someUrls",
      expires: new Date()
    }
  ];

  const component = renderer
    .create(<Spaces key={spaceObj.id} space={spaceObj} />)
    .toJSON();
  expect(component).toMatchSnapshot();
});
