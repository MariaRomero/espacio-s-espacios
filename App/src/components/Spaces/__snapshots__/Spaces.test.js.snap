// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Spaces created correctly 1`] = `
<section
  className="container"
>
  <div
    className="wrapper"
  >
    <ul
      className="list"
    >
      Loading...
    </ul>
  </div>
</section>
`;

exports[`renders components and displays text 1`] = `
<section
  className="container"
>
  <div
    className="wrapper"
  >
    <ul
      className="list"
    >
      Loading...
    </ul>
  </div>
</section>
`;
